cutter/cutter/youtube.py

```diff
@@ -0,0 +1,107 @@
+
+import logging
+import time
+
+import gevent
+import requests
+
+
+class Youtube(object):
+	"""Manages access to youtube and maintains an active access token"""
+
+	ACCESS_TOKEN_ERROR_RETRY_INTERVAL = 10
+	# Refresh token 10min before it expires (it normally lasts an hour)
+	ACCESS_TOKEN_REFRESH_TIME_BEFORE_EXPIRY = 600
+
+	def __init__(self, client_id, client_secret, refresh_token):
+		self.logger = logging.getLogger(type(self).__name__)
+		self.client_id = client_id
+		self.client_secret = client_secret
+		self.refresh_token = refresh_token
+
+		self._first_get_access_token = gevent.spawn(self.get_access_token)
+
+	@property
+	def access_token(self):
+		"""Blocks if access token unavailable yet"""
+		self._first_get_access_token.join()
+		return self._access_token
+
+	def get_access_token(self):
+		while True:
+			try:
+				start_time = time.time()
+				resp = requests.post('https://www.googleapis.com/oauth2/v4/token', data={
+					'client_id': self.client_id,
+					'client_secret': self.client_secret,
+					'refresh_token': self.refresh_token,
+					'grant_type': 'refresh_token',
+				})
+				resp.raise_for_status()
+				data = resp.json()
+				self._access_token = data['access_token']
+				expires_in = (start_time + data['expires_in']) - time.time()
+				if expires_in < self.ACCESS_TOKEN_REFRESH_TIME_BEFORE_EXPIRY:
+					self.logger.warning("Access token expires in {}s, less than normal leeway time of {}s".format(
+						expires_in, self.ACCESS_TOKEN_REFRESH_TIME_BEFORE_EXPIRY,
+					))
+				gevent.spawn_later(expires_in, self.get_access_token)
```

Comment:
Does a stop in the parent Cutter or TranscodeChecker propagate throw to gevent.spawn_later()?

I might be missing something but the this seems to be called exactly at the expiry time and not the ACCESS_TOKEN_REFRESH_TIME_BEFORE_EXPIRY before the expiry.

Reply:
It does not. However, the extra work is harmless if it does happen, and in any case it almost certainly won't since you're generally only stopping when you're bringing down the entire process. The pending greenlet will not prevent process shutdown.

You're correct, good catch. Fixing.